packages/web-forms/tests/components/OdkWebForm.test.ts

```diff
@@ -151,4 +173,274 @@ describe('OdkWebForm', () => {
+			xmlns:orx="http://openrosa.org/xforms">
+			<h:head>
+				<h:title>Edit (basic)</h:title>
+				<model>
+					<instance>
+						<data id="edit-basic">
+							<a>${initialInputValue}</a>
+						</data>
+					</instance>
+					<bind nodeset="/data/a" type="string" />
+				</model>
+			</h:head>
+			<h:body>
+				<input ref="/data/a" />
+			</h:body>
+		</h:html>`;
+
+		let submittedPayload: MonolithicInstancePayload | null = null;
+		let syncResetResult: HostSubmissionResult;
+		let asyncResetResult: Promise<HostSubmissionResult>;
+
+		beforeEach(() => {
+			submittedPayload = null;
+			syncResetResult = { next: POST_SUBMIT__NEW_INSTANCE };
+			asyncResetResult = Promise.resolve(syncResetResult);
+		});
+
+		type HostSubmissionHandler = (
+			payload: MonolithicInstancePayload
+		) => OptionalAwaitableHostSubmissionResult;
+
+		const postSubmissionResetHandler = (
+			payload: MonolithicInstancePayload
+		): HostSubmissionResult => {
+			submittedPayload = payload;
+
+			return syncResetResult;
+		};
+
+		const asyncPostSubmissionResetHandler = (
+			payload: MonolithicInstancePayload
+		): Promise<HostSubmissionResult> => {
+			submittedPayload = payload;
+
+			return asyncResetResult;
+		};
+
+		const postSubmissionNoopHandler = (payload: MonolithicInstancePayload) => {
+			submittedPayload = payload;
+		};
+
+		const asyncPostSubmissionNoopHandler = (payload: MonolithicInstancePayload) => {
+			submittedPayload = payload;
+
+			return Promise.resolve(null);
+		};
+
+		interface SubmissionHandlerCase {
+			readonly description: string;
+			readonly hostSubmissonHandler: HostSubmissionHandler | null;
+			readonly expectedPostSubmissionValue: AssignedInputValue;
+		}
+
+		it.each<SubmissionHandlerCase>([
+			{
+				description: 'resets form state after submission (sync host result)',
+				hostSubmissonHandler: postSubmissionResetHandler,
+				expectedPostSubmissionValue: initialInputValue,
+			},
+			{
+				description: 'resets form state after submission (async host result)',
+				hostSubmissonHandler: asyncPostSubmissionResetHandler,
+				expectedPostSubmissionValue: initialInputValue,
+			},
+			{
+				description: 'does not reset form state by default (sync callback)',
+				hostSubmissonHandler: postSubmissionNoopHandler,
+				expectedPostSubmissionValue: firstSubmissionInputValue,
+			},
+			{
+				description: 'does not reset form state by default (async callback)',
+				hostSubmissonHandler: asyncPostSubmissionNoopHandler,
+				expectedPostSubmissionValue: firstSubmissionInputValue,
+			},
+
+			// EVERYTHING is optional. This case ensures that introducing the callback
+			// as a second parameter doesn't introduce regressions in a host
+			// integration which only uses the first parameter.
+			{
+				description: 'does not reset form state by default (no callback)',
+				hostSubmissonHandler: null,
+				expectedPostSubmissionValue: firstSubmissionInputValue,
+			},
```

Comment:
One thing not tested here, and not currently handled: we should probably expire the callback. I did a very quick spike on having it expire after the current event loop tick (in other words: a host can call the callback immediately with a `Promise`, but the side effect isn't triggered if a host `await`s a `Promise` and then calls the callback after resolution). This is probably solvable, but I wasn't able to do it as trivially as I thought I could. JS event loop scheduling is a lot of things, but it's definitely not _trivial_ (especially a couple weeks into mental fatigue)! 🙃